TypeScript_ESLint.snb.md

```diff
@@ -0,0 +1,189 @@
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts
+
+Note that rules declared in the TypeScript ESLint project use a `util.createRule` helper that internally calls to an `ESLintUtils.RuleCreator`:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/eslint-plugin/src/util/createRule.ts
+
+`RuleCreator` adds a documentation URL to rules under `meta.docs.url` based on their name.
+TypeScript ESLint rules are each documented on a page under https://typescript-eslint.io/rules, such as https://typescript-eslint.io/rules/no-extra-non-null-assertion.
+
+### Rule Metadata Properties
+
+Two particularly useful rule metadata properties are specified in many TypeScript ESLint rules:
+
+- `fixable`: If the rule includes an auto-fixer on any of its messages, either `"code"` or `"whitespace"`
+- `messages`: An object keying ID strings to full messages that may be reported by the rule
+
+`no-extra-non-null-assertion`'s `meta` indicates it has an auto-fixer for code issues and that it may report _`Forbidden extra non-null assertion`_ if it detects an issue.
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L6-17
+
+### Rule Creator Functions
+
+A rule's `create` function returns an object whose keys are AST queries and whose values are functions that will be run whenever a node matching the corresponding query is found.
+The AST query keys use the [ESQuery](https://github.com/estools/esquery) library, which allows using syntax similar to CSS selectors for querying types of nodes.
+
+`no-extra-non-null-assertion` creates three node queries in its returned object:
+
+1. Any `!` non-null assertion within another `!` non-null assertion: e.g. `scream!!.length`
+2. Any `!` non-null assertion just before an `?.` optional chained property: e.g. `scream!?.length`
+3. Any `!` non-null assertion just before an `?.` optional chained call: e.g. `scream!?.()`
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L19-39
+
+Note that any AST node names prefixed with `TS`, including `TSNonNullExpression`, are TypeScript-specific nodes.
+These are available to ESLint rules when using [`@typescript-eslint/parser`](http://npmjs.com/package/@typescript-eslint/parser) instead of ESLint's built-in parser.
+The `@typescript-eslint/utils` package exports those types under the `TSESTree` namespace.
+
+The corresponding TypeScript types for all the possible node types are stored in a [`@typescript-eslint/types`](https://npmjs.com/package/@typescript-eslint/types) package.
```

Comment:
Correction: the types package includes types for all AST nodes as parsed by the parser -- not just TS-specific ones. Existing node types are different now, such as allowing new types of nodes (the TS ones) as children.

```suggestion

For convenience, the `@typescript-eslint/utils` package exports types representing AST nodes as parsed by `@typescript-eslint/parser` under the `TSESTree` namespace.
The corresponding TypeScript types for all the possible node types are stored in a [`@typescript-eslint/types`](https://npmjs.com/package/@typescript-eslint/types) package.
```